src/Symfony/Component/Security/Http/Firewall/ContextListener.php

```diff
@@ -169,21 +170,22 @@ protected function refreshUser(TokenInterface $token)
 
             try {
                 $refreshedUser = $provider->refreshUser($user);
-                $token->setUser($refreshedUser);
+                $newToken = unserialize(serialize($token));
+                $newToken->setUser($refreshedUser);
 
                 // tokens can be deauthenticated if the user has been changed.
-                if (!$token->isAuthenticated()) {
+                if (!$newToken->isAuthenticated()) {
                     if ($this->logoutOnUserChange) {
-                        if (null !== $this->logger) {
-                            $this->logger->debug('Token was deauthenticated after trying to refresh it.', array('username' => $refreshedUser->getUsername(), 'provider' => \get_class($provider)));
-                        }
+                        $deauthenticationContext = array('username' => $refreshedUser->getUsername(), 'provider' => \get_class($provider));
```

Comment:
Would it make sense to collect these data from all providers that reach this point instead of overwriting it every time?